### ISSUE TITLE
Why is Quic so much slower (half the tcp msg/sec)?

### ISSUE THREAD

**Issue by marvin-hansen** [iggy-rs/iggy#606]
Hi,

I ran the benchmark today and I got the following numbers:

TCP:
Results: total throughput: 
402.47 MB/s, 
10061816 messages/s, 
average latency: 0.99 ms, 
average  throughput: 40.25 MB/s, 
total duration: 0.99 s

HTTP:
 Results: total throughput: 
344.18 MB/s, 
8604598 messages/s, 
average latency: 1.16 ms, 
average  throughput: 34.42 MB/s, 
total duration: 1.16 s

QUIC:
 Results: total throughput: 
202.72 MB/s, 
5068119 messages/s, 
average latency: 1.97 ms, 
average  throughput: 20.27 MB/s, 
total duration: 1.97 s

Command:

```bash
RUSTFLAGS='-C target-cpu=native' cargo r --bin iggy-bench -r -- -c -v send --message-size 40 tcp 
```

A few more details:
* My max message size is 40 bytes (SBE Encoded)
* The cpu=native flag usually gives a notch better optimized code

I understand that QUIC is build on top of UDP, but afaik, UDP messaging is usually faster than TCP
because of lower protocol overhead, so it's baffling to see QUIC being only half as fast and 
I wonder if there is any particular reason for that?

Also, it seems that Iggy maxes out at about 20million msg/sec on my box
using the following benchmark:

```bash
RUSTFLAGS='-C target-cpu=native' cargo r --bin iggy-bench -r -- -c -v send-and-poll  --message-size 40 tcp 
```

And the following entry in the Cargo.toml 
```toml
# Optimize all crates
[profile.release]
opt-level = 3
strip = true  # Automatically strip debug symbols from the binary
lto = true    # Enable Link Time Optimization (LTO)
codegen-units = 1 # Reduce Parallel Code Generation Units to Increase Optimization

# There's a Cargo feature named profile-overrides
# that lets you override the optimization level of dependencies.
# https://docs.rust-embedded.org/book/unsorted/speed-vs-size.html
[profile.release.package."*"]
opt-level = 3
strip = true  
codegen-units = 1 
```

20 million msg/sec via tcp is pretty nuts and already blows several commercial systems out of the water.

By your experience, how close is production throughput to the benchmark results?

**Comment by spetz** [comment]
Hello Marvin,

At first, thanks for the kind worst, we're glad that you've found the message throughput to be so efficient already (although, we haven't even touched some low-level kernel optimizations yet, which should make it even faster in the future) 😄. 

May I ask on what machine and OS you ran the benchmarks? Personally, I've found that, for example, running QUIC on Apple Silicon is much slower than when running this transport or Linux or Windows. There are also a bunch of settings in `server.toml` configuration file, which can be adjusted for the QUIC transport, as we're using [Quinn](https://github.com/quinn-rs/quinn).

Also, keep in mind that the fair comparison would be using TCP TLS, as QUIC already enforces you to use the secured connection (however, for the local purposes we use the so-called dangerous configuration with disabled certificate validation).

We will be certainly looking into this, as it should be at least as fast as TCP.

Speaking of the production throughput, we've not heard yet from our users what are their numbers 😃

**Comment by marvin-hansen** [comment]
Ah, thanks for the clarification.

OS is indeed MacOS 14. Apple Silicon M3. Must be an Apple thing. Nevermind,
I'm not going to use quic anyways, but was just curious what the cause
might be.

Also thanks for the hint at TLS,  which makes indeed sense. From memory,
back in the days Sun Microsystem already added crypto co processors to
their mid range Sparc boxes to offload https & TLS  encryption to speed up
processing secured network traffic.

Thing is, I'm working currently with another message systen and I'm not
particularly content with it so I'm contemplating switching to Iggy and
therefore ran the benchmarks first to get an idea what I'm getting myself
if I were to switch. So far hard numbers look a lot better than expected,
so kudos to your good work.

**Comment by spetz** [comment]
Yes, now it makes sense why you've also experienced a bit slower QUIC than expected. I'm happy to hear that the results turned out to be even better than expected! Feel free to ping me and the rest of the team if you'd like to help us with the integration, we've also got a dedicated [Discord](https://iggy.rs/discord) with lots of interesting discussions. 

Please keep in mind, that Iggy currently runs only as single node, however, we've started experimenting with the clustering feature some time ago already (but it will take some time to implement properly, though).

**Comment by hubcio** [comment]
@marvin-hansen we will easily make it much faster, there's tons of improvements on the way, on both Rust SDK and server side.

nice numbers!

**Comment by marvin-hansen** [comment]
I'm good with the throughput, no worries. Adding clustering later is fine
with me as my project is relatively early too.

I'm giving Iggy a try later next week
after I've completed my current project. I've already looked at the Rust
SDK and there's a lot of things done right so I'm quite positive.

Maybe at a comment in the Readme that Quic on Mac isn't as quickly as on
linux and it's a known issue that may get fixed in the future but doesn't
exist on Win/Linux. Just to let people know so your not getting this
question repeatedly whenever someone stumbles across this.

Let's continue on discord and close this for good.
